qmi/instruments/newport/smc_100cc.py

```diff
@@ -30,3 +44,542 @@ def __init__(self,
+            self._name, error_limit
+        )
+        self.controller_address = controller_address
+        # instrument must be in configuration state to persist the set velocity.
+        if persist:
+            self.reset(controller_address)
+            self.enter_configuration_state(controller_address)
+
+        self._scpi_protocol.write(self._build_command(
+            "FE", error_limit))
+        sleep(self.COMMAND_EXEC_TIME)
+        self._check_error()
+        if persist:
+            self.exit_configuration_state(controller_address)
+
+    @rpc_method
+    def get_friction_compensation(self, controller_address: Optional[int] = None) -> float:
+        """
+        Get the friction compensation.
+
+        Parameters:
+            controller_address: Optional address of the controller that needs to be controlled. By default,
+                                it is set to the initialised value of the controller address.
+        """
+        # instrument must be in configuration state to get the friction compensation.
+        _logger.info(
+            "Getting the friction compensation of instrument [%s]", self._name)
+        self.reset(controller_address)
+        self.enter_configuration_state(controller_address)
+        friction_compensation = self._scpi_protocol.ask(
+            self._build_command("FF?"))
+        self._check_error()
+        self.exit_configuration_state(controller_address)
+        return float(friction_compensation[3:])
+
+    @rpc_method
+    def set_friction_compensation(
+            self, friction_compensation: float, persist: bool = False, controller_address: Optional[int] = None
+    ) -> None:
+        """
+        Set the friction compensation. It must not be larger than the driver voltage set by DV command.
+
+        Parameters:
+            friction_compensation:      New friction compensation value.
+            persist:            Flag to indicate if the friction compensation should be persisted to the controller's
+                                memory, so it is still available after powering down the controller. When not persisted,
+                                the friction compensation is the one stored in the controller's memory.
+            controller_address: Optional address of the controller that needs to be controlled. By default,
+                                it is set to the initialised value of the controller address.
+        """
+        self.controller_address = controller_address
+        response = self._scpi_protocol.ask(
+            self._build_command("DV?"))
+        sleep(self.COMMAND_EXEC_TIME)
+        self._check_error()
+        driver_voltage = float(response[3:])
+        if driver_voltage <= friction_compensation or friction_compensation < 0:
+            raise QMI_InstrumentException(
+                f"Provided value {friction_compensation} not in valid range 0 >= friction_compensation > {driver_voltage}."
+            )
+
+        _logger.info(
```

Comment:
this is a personal preference, but I like these log calls to be the first call in a method

Reply:
Well I think for that the logger messages would have to be written in form "Entering method {method_name} with input parameters {param_1}, {param_2}, ...", as the current log message says "Setting the ...". This is of course not happening at all if the input parameters are not valid, so that's why I moved them behind the input parameter checks in all relevant methods. 

I also moved them after the first command to set the `self.controller_address` if the logger message makes use of it in the message - as taking directly the input parameter `controller_address` could give `None`.

Reply:
Eventually, in my "Final" solution we always set the `self.controller_address = controller_address` before the logging message in all the calls. But after possible parameter checks if no other actions are required.